app/scripts/lib/tx-gas-utils.js

```diff
@@ -37,14 +38,30 @@ module.exports = class txProvideUtil {
 
   async estimateTxGas (txMeta, blockGasLimitHex) {
     const txParams = txMeta.txParams
+
     // check if gasLimit is already specified
     txMeta.gasLimitSpecified = Boolean(txParams.gas)
-    // if not, fallback to block gasLimit
-    if (!txMeta.gasLimitSpecified) {
-      const blockGasLimitBN = hexToBn(blockGasLimitHex)
-      const saferGasLimitBN = BnMultiplyByFraction(blockGasLimitBN, 19, 20)
-      txParams.gas = bnToHex(saferGasLimitBN)
+
+    // if it is, use that value
+    if (txMeta.gasLimitSpecified) {
+      return txParams.gas
     }
+
+    // if recipient has no code, gas is 21k max:
+    const recipient = txParams.to
+    const hasRecipient = Boolean(recipient)
+    const code = await this.query.getCode(recipient)
+    if (hasRecipient && (!code || code === '0x')) {
+      txParams.gas = SIMPLE_GAS_COST
+      txMeta.gasLimitSpecified = true // Prevents buffer addition
```

Comment:
`gasLimitSpecified` is used to suggest that the dapp suggested the `gasLimit`, lets not overload that meaning

Reply:
seems that bc we return here we dont add the buffer addition, and not because of setting the flag

Reply:
No, the buffer is added even if you simply return here.

Reply:
> lets not overload that meaning

I could just add another boolean flag that means the same thing?

Reply:
Whatever is returned from this method is passed to `setTxGas`, which adds a buffer unless the flag is set.

Reply:
ah i see. logic flow is sort of strange here, could use some cleanup. e.g. we set to `19/20 * block.gasLimit` then override that with an `estimateGas`


```js
if (txMeta.gasLimitSpecified || txParams.gas === SIMPLE_GAS_COST) {
  //  ...
}
```